src/main/java/com/epam/aidial/core/service/CustomApplicationService.java

```diff
@@ -0,0 +1,177 @@
+package com.epam.aidial.core.service;
+
+import com.epam.aidial.core.ProxyContext;
+import com.epam.aidial.core.config.Application;
+import com.epam.aidial.core.data.ListSharedResourcesRequest;
+import com.epam.aidial.core.data.MetadataBase;
+import com.epam.aidial.core.data.NodeType;
+import com.epam.aidial.core.data.ResourceFolderMetadata;
+import com.epam.aidial.core.data.ResourceType;
+import com.epam.aidial.core.data.SharedResourcesResponse;
+import com.epam.aidial.core.security.AccessService;
+import com.epam.aidial.core.security.EncryptionService;
+import com.epam.aidial.core.storage.BlobStorageUtil;
+import com.epam.aidial.core.storage.ResourceDescription;
+import com.epam.aidial.core.util.ProxyUtil;
+import io.vertx.core.json.JsonObject;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+@Slf4j
+@RequiredArgsConstructor
+public class CustomApplicationService {
+
+    private static final int FETCH_SIZE = 1000;
+
+    private final EncryptionService encryptionService;
+    private final ResourceService resourceService;
+    private final ShareService shareService;
+    private final AccessService accessService;
+    private final JsonObject applicationsConfig;
+
+    /**
+     * Loads a custom application from provided url with permission check
+     *
+     * @param url - custom application url
+     * @return - custom application or null if invalid url provided or resource not found
+     */
+    public Application getCustomApplication(String url, ProxyContext context) {
+        ResourceDescription resource;
+        try {
+            resource = ResourceDescription.fromAnyUrl(url, encryptionService);
+        } catch (Exception e) {
+            log.warn("Invalid resource url provided: {}", url);
```

Comment:
Why don't you throw an exception?

Reply:
to keep an old behavior, if no deployment found - we return a 404

Reply:
throw new ResourceNotFoundException?

Reply:
fixed